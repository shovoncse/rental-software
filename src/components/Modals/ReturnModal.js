import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Row, Col, Modal, Form } from 'antd';
import ProductDetails from 'components/Others/ProductDetails';
import ConfirmDialogue from 'components/Others/ConfirmDialogue';
import ReturnForm from 'components/Forms/ReturnForm';
import ErrorBoundary from 'utils/ErrorBoundary';
import Swal from 'sweetalert2';
import Button from 'components/Buttons/Button';

const ReturnModal = ({ visible, setReturn }) => {
  let dispatch = useDispatch();
  let selected = useSelector((state) => state.products.selected);

  const [returnForm] = Form.useForm();
  const [confirm, setConfirm] = useState(false);
  // Returning product
  const handleReturnOk = ({ mileage }) => {
    dispatch({ type: 'products/updateSelected', payload: { mileage } });
    setConfirm(true);
  };

  useEffect(() => {
    if (!visible) {
      returnForm.resetFields();
      dispatch({ type: 'products/resetSelected' });
    }
  }, [visible, returnForm, dispatch]);
  // Confirm
  const handleReturnConfirm = () => {
    dispatch({ type: 'products/returnProduct' });
    Swal.fire('Sucessfull', 'Your Return is successfull', 'success');
    setConfirm(false);
    setReturn(false);
  };

  const handleReturn = () => {
    returnForm
      .validateFields()
      .then((values) => {
        handleReturnOk(values);
      })
      .catch((info) => {
        console.log('Validate Failed:', info);
      });
  };

  return (
    <>
      <Modal
        visible={visible}
        title="Return a product"
        okText="Yes"
        cancelText="no"
        onCancel={confirm ? () => setConfirm(false) : () => setReturn(false)}
        footer={[
          <Button
            type="red"
            key={1}
            text="No"
            click={confirm ? () => setConfirm(false) : () => setReturn(false)}
          />,
          <Button
            type="blue"
            key={2}
            text="Yes"
            click={confirm ? () => handleReturnConfirm() : () => handleReturn()}
          />,
        ]}
      >
        {!confirm ? (
          <Row justify="space-between">
            <Col span={10}>
              {/* Return Form */}
              <ErrorBoundary>
                <ReturnForm name={returnForm} />
              </ErrorBoundary>
            </Col>
            <Col span={12}>
              {selected ? (
                <ErrorBoundary>
                  <ProductDetails selected={selected} />
                </ErrorBoundary>
              ) : null}
            </Col>
          </Row>
        ) : (
          <ErrorBoundary>
            <ConfirmDialogue selected={selected} />
          </ErrorBoundary>
        )}
      </Modal>
    </>
  );
};

export default ReturnModal;
