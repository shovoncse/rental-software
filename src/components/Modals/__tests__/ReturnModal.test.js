import React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import ReturnModal from '../ReturnModal';
import { store } from 'store/';
import { Provider } from 'react-redux';
afterEach(cleanup);

describe('<ReturnModal />', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Provider store={store}>
        <ReturnModal></ReturnModal>
      </Provider>,
      div
    );
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ReturnModal></ReturnModal>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
