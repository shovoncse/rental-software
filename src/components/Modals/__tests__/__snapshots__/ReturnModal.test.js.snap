// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReturnModal /> matches snapshot 1`] = `null`;
